### PR TITLE
DRAFT: support multiple versions in same hierarchy

### PR DESCRIPTION
**Pull request by goyal1092**
## Checklist

- [ ]  🚀 is the code ready to be merged and go live?
- [ ]  🛠 does it work (build) locally

### Pull Request

- [ ]  📰 good title
- [ ]  📝good description
- [ ]  🔖 issue linked
- [ ]  📖 changelog filled out

### Commits

- [ ]  commits are clean
- [ ]  commit messages are clean

### Code Quality

- [ ]  🚧 no commented out code
- [ ]  🖨 no unnecessary logging
- [ ]  🎱 no magic numbers
- [ ]  black was run locally (as part of the pre-commit hook)

### Testing

- [ ]  ✅ added (appropriate) unit tests
- [ ]  💢 edge cases in tests were considered
- [ ]  ✅ ran tests locally & are passing

### CODE REVIEW

File: wazimap_ng/profile/serializers/metrics_serializer.py [hunk 1]
```diff
@@ -16,20 +16,20 @@ def get_data_for_key_metric_and_geographies(profile_key_metric, geographies):
...
-def subindicator(profile_key_metric, geography):
+def subindicator(profile_key_metric, geography, versions):
     indicator_data = get_data_for_key_metric_and_geographies(profile_key_metric, [geography])
     if indicator_data != None:
         return MetricCalculator.subindicator(indicator_data, profile_key_metric, geography)
     return None
 
-def sibling(profile_key_metric, geography):
-    siblings = geography.get_siblings()
+def sibling(profile_key_metric, geography, versions):
```
Comment (adieyal) [inline wazimap_ng/profile/serializers/metrics_serializer.py:31]: I don't think this is the right thing to do here. The sibling calculate looks at the % that this geography represents as a function of all the surrounding geographies - e.g. CPT contains 50% of the population in WC

File: wazimap_ng/urls.py [hunk 2]
```diff
@@ -149,6 +149,11 @@ def trigger_error(request):
         name="all-details"
     ),
 
+    path(
+        "api/v1/indicator_data/profile/<int:profile_id>/geography/<str:geography_code>/",
+        cache(general_views.consolidated_versioned_data),
+        name="all-details"
```
Comment (adieyal) [inline wazimap_ng/urls.py:155]: this has the same name as the previous url

File: wazimap_ng/urls.py [hunk 3]
```diff
@@ -149,6 +149,11 @@ def trigger_error(request):
         name="all-details"
     ),
 
+    path(
+        "api/v1/indicator_data/profile/<int:profile_id>/geography/<str:geography_code>/",
```
Comment (adieyal) [inline wazimap_ng/urls.py:153]: does this return all the data for a single geography_code? How is it different to all-details?

**Review by adieyal — CHANGES_REQUESTED** [review]

File: wazimap_ng/urls.py (hunk 2 above)
Comment (goyal1092) [reply wazimap_ng/urls.py:155]: Will change that right now

File: wazimap_ng/urls.py (hunk 3 above)
Comment (goyal1092) [reply wazimap_ng/urls.py:153]: Not specified in url but format should be `Geography_code:version1:version2:version3`
Will look for proper way to make this type of url.
In backend we are splitting geography_code by `:` and get list of versions and geography code

File: wazimap_ng/boundaries/management/commands/loadshp.py [hunk 4]
```diff
@@ -33,28 +33,34 @@ def check_shapefile(self, shapefile_path):
...
         try:
-            parent_geography = Geography.objects.get(code__iexact=parent_code, version=version)
+            parent_geography = Geography.objects.get(code__iexact=parent_code, versions=version)
         except Geography.DoesNotExist:
             print(f"Can't find parent geography: {parent_code} ({version})")
             return
 
         try:
-            geography = Geography.objects.get(code__iexact=fields["code"], version=version)
+            geography = Geography.objects.get(code__iexact=fields["code"], versions=version)
```
Comment (jbothma) [inline wazimap_ng/boundaries/management/commands/loadshp.py:58]: `versions=version` reads weirdly

File: wazimap_ng/boundaries/models.py [hunk 5]
```diff
@@ -1,25 +1,22 @@
...
+    version = models.ForeignKey(Version, on_delete=models.PROTECT, null=True)
     area = models.FloatField()
     geom = models.MultiPolygonField(srid=4326, null=True)
     geom_cache = CachedMultiPolygonField(field_name="geom")
-    objects = GeographyBoundaryManager()
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(fields=["version", "geography"], name="unique_geography_version")
+        ]
```
Comment (jbothma) [inline wazimap_ng/boundaries/models.py:21]: unique_together seems to be a built-in way to do this https://docs.djangoproject.com/en/3.2/ref/models/options/#unique-together, or is there a benefit to this more explicit form?

File: wazimap_ng/boundaries/views.py [hunk 6]
```diff
@@ -33,24 +35,29 @@ def get_geography(self, code):
 
 @cache_decorator("geography_item")
 def geography_item_helper(code, version):
-    geography = get_object_or_404(Geography, code=code, version=version)
-    serializer = serializers.GeographyBoundarySerializer(geography.geographyboundary)
+    geography = get_object_or_404(Geography, code=code)
+    boundary = geography.geographyboundary_set.filter(version=version).first()
```
Comment (jbothma) [inline wazimap_ng/boundaries/views.py:39]: Is this relying on boundary geography and version pair to be unique together? Perhaps worth a comment.
Alternatively
Boundary.get(version=version, geography=geography) which would raise if there's more than one

File: wazimap_ng/boundaries/views.py [hunk 7]
```diff
@@ -33,24 +35,29 @@ def get_geography(self, code):
...
 
 @cache_decorator("geography_children")
 def geography_children_helper(code, version):
-
-    geography = Geography.objects.get(code=code, version=version)
-    child_boundaries = geography.get_child_boundaries()
-    children = geography.get_children()
+    geography = Geography.objects.get(code=code)
+    child_boundaries = geography.get_child_boundaries(version)
+    children = geography.get_child_geographies(version)
```
Comment (jbothma) [inline wazimap_ng/boundaries/views.py:50]: is this tested?

File: wazimap_ng/datasets/admin/forms/dataset_admin_form.py [hunk 8]
```diff
@@ -34,9 +33,4 @@ def __init__(self, *args, **kwargs):
...
 
             if filtered_profile:
-                self.fields["profile"].initial = filtered_profile
-
-                for hierarchy in hierarchies:
-                    hp = hierarchy.profile_set.all()
-                    if hp.count() == 1 and hp.first() == filtered_profile:
-                        self.fields["geography_hierarchy"].initial = hierarchy
\ No newline at end of file
+                self.fields["profile"].initial = filtered_profile
```
Comment (jbothma) [inline wazimap_ng/datasets/admin/forms/dataset_admin_form.py]: is this tested?

File: wazimap_ng/datasets/models/dataset.py [hunk 9]
```diff
@@ -22,9 +22,9 @@ class Dataset(BaseModel):
     )
     name = models.CharField(max_length=60)
     groups = ArrayField(models.CharField(max_length=200), blank=True, default=list)
-    geography_hierarchy = models.ForeignKey(GeographyHierarchy, on_delete=models.CASCADE)
     permission_type = models.CharField(choices=PERMISSION_TYPES, max_length=32, default="private")
     content_type = models.CharField(choices=DATASET_CONTENT_TYPES, max_length=32, default=QUANTITATIVE)
+    version = models.ForeignKey(Version, on_delete=models.CASCADE, null=True)
```
Comment (jbothma) [inline wazimap_ng/datasets/models/dataset.py:27]: probably shouldn't be null=True

File: wazimap_ng/general/views.py [hunk 10]
```diff
@@ -52,17 +58,43 @@ def consolidated_profile_helper(profile_id, geography_code):
...
+    versions_list = profile.geography_hierarchy.configuration.get("versions", [])
+    versions = list(set(versions_list).intersection(versions))
+    version_objs = list(dataset_models.Version.objects.filter(name__in=versions))
+
+    geography = dataset_models.Geography.objects.get(code=geo_code)
+
+    if not version_objs or not geography:
+        raise Http404
+    js = profile_serializers.ExtendedProfileSerializer(profile, geography, version_objs)
+    return Response(js)
```
Comment (jbothma) [inline wazimap_ng/general/views.py:97]: probably wanna keep this out of the first iteration on this

File: wazimap_ng/urls.py (hunk 3 above)
Comment (jbothma) [reply wazimap_ng/urls.py:153]: Probably wanna keep this out of the first iteration on this

**Comment by jbothma** [comment]
Replaced with https://github.com/OpenUpSA/wazimap-ng/pull/341 where we can all push
